types/src/v0/impls/state.rs

```diff
@@ -253,130 +303,329 @@ impl ValidatedState {
+        if self.header.l1_head() < parent_l1_head {
+            return Err(ProposalValidationError::DecrementingL1Head);
         }
+        Ok(())
+    }
+    /// Validates [`ChainConfig`] of proposal against that stored in
+    /// state by comparing commitments.
+    fn validate_chain_config(
+        &self,
+        expected_chain_config: &ChainConfig,
+    ) -> Result<(), ProposalValidationError> {
+        let proposed_chain_config = self.header.chain_config();
+        if proposed_chain_config.commit() != expected_chain_config.commit() {
+            return Err(ProposalValidationError::InvalidChainConfig {
+                expected: Box::new(*expected_chain_config),
+                proposal: Box::new(proposed_chain_config),
+            });
+        }
+        Ok(())
     }
-}
 
-impl From<NsTableValidationError> for ProposalValidationError {
-    fn from(err: NsTableValidationError) -> Self {
-        Self::InvalidNsTable { err }
+    /// Validate timestamp is not decreasing relative to parent.
+    fn validate_timestamp_non_dec(
+        &self,
+        parent_timestamp: u64,
+    ) -> Result<(), ProposalValidationError> {
+        if self.header.timestamp() < parent_timestamp {
+            return Err(ProposalValidationError::DecrementingTimestamp {
+                proposal_timestamp: self.header.timestamp(),
+                parent_timestamp,
+            });
+        }
+
+        Ok(())
     }
-}
 
-pub fn validate_proposal(
-    state: &ValidatedState,
+    /// Validate timestamp is within a given tolerance of system
+    /// time. Tolerance is currently `12` seconds. This value may be
+    /// moved to configuration in the future.
+    fn validate_timestamp_drift(&self, system_time: u64) -> Result<(), ProposalValidationError> {
+        // TODO 12 seconds of tolerance should be enough for reasonably
+        // configured nodes, but we should make this configurable.
+        let diff = self.header.timestamp().abs_diff(system_time);
+        if diff > 12 {
+            return Err(ProposalValidationError::InvalidTimestampDrift {
+                proposal: self.header.timestamp(),
+                system: system_time,
+                diff,
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Validate Block Merkle Tree by comparing proposed commitment against expectation.
+    fn validate_block_merkle_tree(
+        &self,
+        block_merkle_tree_root: BlockMerkleCommitment,
+    ) -> Result<(), ProposalValidationError> {
+        if self.header.block_merkle_tree_root() != block_merkle_tree_root {
+            return Err(ProposalValidationError::InvalidBlockRoot {
+                expected_root: block_merkle_tree_root,
+                proposal_root: self.header.block_merkle_tree_root(),
+            });
+        }
+
+        Ok(())
+    }
+}
+/// Type to hold cloned validated state and provide validation methods.
+#[derive(Debug)]
+pub(crate) struct ValidatedTransition<'a> {
```

Comment:
A bit of a nitpick: the transition is only validated if `validate()` does not return an error but given current naming one might expect it to be validated upon creation unless one actually inspects the return type. I'm totally fine with the current name just thinking maybe we can make it even clearer at some point.

Reply:
Its a tough one. It's a type that only exists to facilitate a state transition by encapsulating validation logic. The state is actually held in `ValidatedState`. Open to ideas, but I don't have any at the time being.